### PR TITLE
drivers: lora: reyax rylr module

### PR DESCRIPTION
**Pull request by DaWei8823**
writing a driver that implements the lora api for  [reyax lora rylr series modules](https://reyax.com/products/RYLR896) using the zephyr modem utilities.

Note: the bulk of this driver is implemented in `drivers/lora/rylrxxx.c` which must be opened manually (marked by "Large diffs are not rendered by default" warning) 

I tested with two modules: one connected to my board and the other connected to my PC via uart adapter. 

I built the `samples/hello_world` project
```
west build -b cy8ckit_062s4 --pristine -- -DSHIELD=reyax_lora -DCONFIG_LOG=y -DCONFIG_SHELL=y -DCONFIG_LORA_SHELL=y
west flash
```

in the zephyr shell, I sent:
```
uart:~$ lora config pre-len 7
uart:~$ lora config freq 865100000
uart:~$ lora send HELLO
```

on the serial console in my PC, I saw:
```
+RCV=1,5,HELLO,-41,36
```

then I ran this command in the zephyr terminal:
```
uart:~$ lora recv 
```
and I sent this command to the PC-connected module:
```
AT+SEND=0,3,sup
```
I saw this output in zephyr terminal:

```
00000000: 73 75 70                                         |sup              |
RSSI: 208 dBm, SNR:38 dBm
```

Signed-off-by: David Ullmann <davidl.ullmann@gmail.com>

### CODE REVIEW

File: drivers/lora/CMakeLists.txt [hunk 1]
```diff
@@ -20,3 +20,6 @@ endif()
 
 zephyr_library_sources_ifdef(CONFIG_LORA_SX126X sx126x_standalone.c)
 zephyr_library_sources_ifdef(CONFIG_LORA_STM32WL_SUBGHZ_RADIO sx126x_stm32wl.c)
+zephyr_library_sources_ifdef(CONFIG_LORA_RYLRXXX rylrxxx.c)
+
+zephyr_library_include_directories_ifdef(CONFIG_LORA_RYLRXXX ${ZEPHYR_BASE}/drivers/modem)
```
Comment (JordanYates) [inline drivers/lora/CMakeLists.txt:25]: This shouldn't be required. If there is a header required externally, it should be under `${ZEPHYR_BASE}/include/drivers/modem`

File: drivers/lora/Kconfig.rylrxxx [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+	depends on DT_HAS_REYAX_RYLRXXX_ENABLED
+	select MODEM
+	select MODEM_CONTEXT
+	select MODEM_CMD_HANDLER
+	select MODEM_IFACE_UART
+	select UART_USE_RUNTIME_CONFIGURE
+	help
+		Enable the Reyax LYLR driver
+
+config LORA_RYLRXX_RX_POOL_COUNT
```
Comment (JordanYates) [inline drivers/lora/Kconfig.rylrxxx:19]: The options from here down should be guarded in a block like so:
```
if LORA_RYLRXXX

options go here...

endif # LORA_RYLRXXX
```
This prevents the options being configured when the modem isn't enabled

File: drivers/lora/Kconfig.rylrxxx [hunk 3]
```diff
@@ -0,0 +1,48 @@
+#
+# Copyright (c) 2024 David Ullmann
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+config LORA_RYLRXXX
+	bool "Reyax LYLR driver"
+	default n
```
Comment (JordanYates) [inline drivers/lora/Kconfig.rylrxxx:9]: `default n` is never required on a config symbol

File: drivers/lora/Kconfig.rylrxxx [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+	depends on DT_HAS_REYAX_RYLRXXX_ENABLED
+	select MODEM
+	select MODEM_CONTEXT
+	select MODEM_CMD_HANDLER
+	select MODEM_IFACE_UART
+	select UART_USE_RUNTIME_CONFIGURE
+	help
+		Enable the Reyax LYLR driver
```
Comment (JordanYates) [inline drivers/lora/Kconfig.rylrxxx:17]: help text should be indented by 2 spaces from the rest of the text, take a look at the other upstream Kconfig files for examples.

File: boards/arm/cy8ckit_062s4/cy8ckit_062s4_m4_rylr.overlay
Comment (JordanYates) [inline boards/arm/cy8ckit_062s4/cy8ckit_062s4_m4_rylr.overlay:1]: This shouldn't be an overlay at the board level, as I assume this is specific to how you personally connected the module to the dev kit. This should be a Zephyr shield https://docs.zephyrproject.org/latest/hardware/porting/shields.html.

From what I can tell the dev board board uses the Arduino pin layout for headers, so the device should probably sit under an `arduino_serial` node. See `nrf52840dk_nrf52840.dts` for an example.

File: drivers/lora/rylrxxx.c [hunk 5]
```diff
@@ -0,0 +1,632 @@
...
+
+#define RYLR_DEVICE_INIT(n)                                                                        \
+	NET_BUF_POOL_DEFINE(rylr_recv_pool_##n, CONFIG_LORA_RYLRXX_RX_POOL_COUNT,                  \
+			    RYLR_MAX_RESPONSE, 0, NULL);                                           \
+	static K_KERNEL_STACK_DEFINE(modem_rx_stack_##n, CONFIG_RYLRXX_RX_STACK_SIZE);             \
+	K_MSGQ_DEFINE(recv_msgq_##n, sizeof(struct rylr_recv_msg),                                 \
+		      CONFIG_RYLRXXX_UNSOLICITED_RX_MSGQ_SIZE, sizeof(uint32_t));                  \
+	K_SEM_DEFINE(rylr_resp_sem_##n, 0, 1);                                                     \
+	K_SEM_DEFINE(rylr_operation_sem_##n, 1, 1);                                                \
+	static struct rylr_data dev_data_##n = {                                                   \
```
Comment (JordanYates) [inline drivers/lora/rylrxxx.c:618]: Unless you have a good reason the `data` struct shouldn't be initialised to anything, it should be setup in the `init` function. If you are trying to point to other variables that need to be defined, those pointers should be part of the config struct.

**Review by JordanYates — CHANGES_REQUESTED** [review]
Quick review without examining the substance of the driver implementation

**Comment by bjarki-trackunit** [comment]
Hi, new modems should not use the old modem libraries like `modem_cmd_handler` and `modem_iface_uart`. See https://github.com/zephyrproject-rtos/zephyr/tree/main/include/zephyr/modem and https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/gnss/gnss_quectel_lcx6g.c :)

File: drivers/lora/rylrxxx.c (hunk 5 above)
Comment (JordanYates) [reply drivers/lora/rylrxxx.c:618]: For variables in the data struct yes. The problem with static init of RAM variables is that it then requires a copy of that entire struct in ROM as well to initialise from. Given the large size of the data struct in this instance, that's a lot of overhead.

**Comment by DaWei8823** [comment]
@JordanYates and @bjarki-trackunit, thank you for taking a look earlier!  I have addressed your comments and updated the PR

**Comment by DaWei8823** [comment]
@Mani-Sadhasivam @ifyall @galak @nashif could you take a look at your earliest convenience? thanks!

**Comment by DaWei8823** [comment]
@cfriedt could you take a look when you get a chance?

**Review by JordanYates — CHANGES_REQUESTED** [review]
Can you please split the commits into:
1. Vendor prefix
2. LoRa changes
3. Board changes
4. Shield implementation

I will need to defer to someone else for the modem chat implementation, I haven't used that API and I don't have hardware to test with.

**Comment by DaWei8823** [comment]
@JordanYates just wanted to follow up on this!

File: boards/shields/reyax_lora/reyax_lora.overlay [hunk 6]
```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright (c) 2024 David Ullmann
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/
```
Comment (JordanYates) [inline boards/shields/reyax_lora/reyax_lora.overlay:6]: extra `/`?

File: drivers/lora/rylrxxx.c [hunk 7]
```diff
@@ -0,0 +1,684 @@
...
+		return err;
+	}
+
+	err = modem_pipe_open(data->uart_pipe);
+	if (err != 0) {
+		LOG_ERR("error opening uart pipe %d", err);
+		return err;
+	}
+
+	err = gpio_pin_set_dt(&config->reset, 1);
```
Comment (JordanYates) [inline drivers/lora/rylrxxx.c:634]: Reset line is already being set to active by the `GPIO_OUTPUT_ACTIVE` flag in `gpio_pin_configure_dt` above.

**Review by JordanYates — CHANGES_REQUESTED** [review]
Only thing missing now is to ensure this driver is built by CI in some form.
Ideally `samples/drivers/lora/send` and `samples/drivers/lora/receive`

**Comment by cfriedt** [comment]
@DaWei8823 - def want to build the driver in CI as Jordan suggested. The sample application sounds like a good spot.

**Comment by DaWei8823** [comment]
> Only thing missing now is to ensure this driver is built by CI in some form. Ideally `samples/drivers/lora/send` and `samples/drivers/lora/receive`

@JordanYates, can you please point me to some resources for adding a CI job that would do something like `west build -b cy8ckit_062s4 --pristine -- -DSHIELD=reyax_lora -DCONFIG_LOG=y -DCONFIG_SHELL=y -DCONFIG_LORA_SHELL=y
`

**Comment by JordanYates** [comment]
> > Only thing missing now is to ensure this driver is built by CI in some form. Ideally `samples/drivers/lora/send` and `samples/drivers/lora/receive`
> 
> @JordanYates, can you please point me to some resources for adding a CI job that would do something like `west build -b cy8ckit_062s4 --pristine -- -DSHIELD=reyax_lora -DCONFIG_LOG=y -DCONFIG_SHELL=y -DCONFIG_LORA_SHELL=y `

Something like this: https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/build_all/rtc/testcase.yaml
```
  sample.driver.lora.rylr.send:
    extra_args: SHIELD=the_shield
    platform_allow: the_platform
```
You should be able to validate whether it works locally with `zephyr/scripts/twister --all -T zephyr/samples/drivers/lora/send`. You might need to move the `depends_on: lora` part from common, I'm not sure how `depends_on` interacts with shields.

File: drivers/lora/rylrxxx.c (hunk 7 above)
Comment (DaWei8823) [reply drivers/lora/rylrxxx.c:634]: ahh right, I'm actually going to take  GPIO_OUTPUT_ACTIVE out of gpio_pin_configure_dt as it feels saner to me for pin configuration and activation to be two different steps

File: boards/shields/reyax_lora/reyax_lora.overlay (hunk 6 above)
Comment (DaWei8823) [reply boards/shields/reyax_lora/reyax_lora.overlay:6]: ahh it just belongs after the #include (but must go before the root node)

**Comment by DaWei8823** [comment]
@cfriedt, @JordanYates I've made the requested changes. Thanks for your comments! Please take a look when you can

**Review by cfriedt — DISMISSED** [review]
LGTM

**Comment by DaWei8823** [comment]
@JordanYates, just following up on this

**Review by JordanYates — DISMISSED** [review]

File: boards/shields/reyax_lora/doc/index.rst [hunk 8]
```diff
@@ -0,0 +1,29 @@
+.. lora_reyax:
+
+Reyax LoRa RYLR896 and RYLR915 Modules
+############################################################################
```
Comment (aescolar) [inline boards/shields/reyax_lora/doc/index.rst:4]: Nit: A few "#" too many :)

**Review by aescolar — DISMISSED** [review]
Please add an entry in the release notes mentioning this new driver

**Review by mniestroj — DISMISSED** [review]

File: boards/shields/reyax_lora/doc/index.rst (hunk 8 above)
Comment (DaWei8823) [reply boards/shields/reyax_lora/doc/index.rst]: @aescolar done

**Comment by DaWei8823** [comment]
> Please add an entry in the release notes mentioning this new driver

@aescolar done

File: drivers/lora/rylrxxx.c [hunk 9]
```diff
@@ -0,0 +1,684 @@
...
+	data->dynamic_chat.response_matches = &ok_match;
+	data->dynamic_chat.response_matches_size = 1;
+	data->dynamic_chat.timeout = 0;
+
+	data->dynamic_script.script_chats = &data->dynamic_chat;
+	data->dynamic_script.script_chats_size = 1;
+	data->dynamic_script.abort_matches = abort_matches;
+	data->dynamic_script.abort_matches_size = 1;
+	data->dynamic_script.callback = on_script_finished;
+	data->dynamic_script.timeout = CONFIG_RYLRXXX_RADIO_CMD_RESPONSE_TIMEOUT_MS;
```
Comment (cfriedt) [inline drivers/lora/rylrxxx.c:218]: Not blocking or anything, just a suggestion
```suggestion
	data->dynamic_chat = (<cast to type of dynamic chat>) {
		.response_matches = &ok_match;
		.response_matches_size = 1;
		.script_chats = &data->dynamic_chat;
		.script_chats_size = 1;
		.abort_matches = abort_matches;
		.abort_matches_size = 1;
		.callback = on_script_finished;
		.timeout = CONFIG_RYLRXXX_RADIO_CMD_RESPONSE_TIMEOUT_MS;
};
```

**Review by cfriedt — DISMISSED** [review]

File: boards/shields/reyax_lora/doc/index.rst (hunk 8 above)
Comment (aescolar) [reply boards/shields/reyax_lora/doc/index.rst]:
```suggestion
* LoRa

  * Added driver for Reyax LoRa module

```
You need an extra linebreak before and after so it renders properly 
https://builds.zephyrproject.io/zephyr/pr/69273/docs/releases/release-notes-3.7.html#drivers-and-sensors

File: doc/releases/release-notes-3.7.rst [hunk 10]
```diff
@@ -188,6 +188,8 @@ Drivers and Sensors
 
 * Input
 
+* LoRa
+  * Added driver for Reyax LoRa module
```
Comment (aescolar) [inline doc/releases/release-notes-3.7.rst:192]:
```suggestion
* LoRa

  * Added driver for Reyax LoRa module

```
You need an extra linebreak before and after so it renders properly 
https://builds.zephyrproject.io/zephyr/pr/69273/docs/releases/release-notes-3.7.html#drivers-and-sensors

Comment (DaWei8823) [reply doc/releases/release-notes-3.7.rst]: > You need an extra linebreak before and after so it renders properly https://builds.zephyrproject.io/zephyr/pr/69273/docs/releases/release-notes-3.7.html#drivers-and-sensors

done: https://github.com/zephyrproject-rtos/zephyr/blob/1a7cabc81eb467390d7262c29bfcce6b3394c2df/doc/releases/release-notes-3.7.rst

File: boards/shields/reyax_lora/reyax_lora.overlay [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+
+&arduino_serial {
+    current-speed = <115200>;
+    status = "okay";
+    rylr_lora_modem: rylr_lora_modem {
+            compatible = "reyax,rylrxxx";
+            status = "okay";
+            reset-gpios = <&arduino_header 16 GPIO_ACTIVE_LOW>;
+    };
+};
```
Comment (mniestroj) [inline boards/shields/reyax_lora/reyax_lora.overlay:23]: Please convert whole file to use tabs for indentation, to be consistent with other files in repo. Same with `boards/shields/reyax_lora/boards/cy8ckit_062s4.overlay`

**Review by mniestroj — COMMENTED** [review]
Nit about whitespaces in DT overlay files.

File: boards/shields/reyax_lora/reyax_lora.overlay (hunk 11 above)
Comment (DaWei8823) [reply boards/shields/reyax_lora/reyax_lora.overlay:23]: @mniestroj done

**Review by mniestroj — DISMISSED** [review]

**Review by JordanYates — DISMISSED** [review]

**Comment by DaWei8823** [comment]
@cfriedt can this be merged now that's it's been approved?

**Comment by cfriedt** [comment]
> @cfriedt can this be merged now that's it's been approved?

@DaWei8823 - it needs a rebase unfortunately because there are merge conflicts. But given all of the approvals, it should be quick after that.

**Review by mniestroj — DISMISSED** [review]

**Review by cfriedt — DISMISSED** [review]

File: boards/infineon/cy8ckit_062s4/cy8ckit_062s4.dts [hunk 12]
```diff
@@ -29,6 +29,32 @@
 
 	};
 
+	arduino_header: connector {
+		compatible = "arduino-header-r3";
+		#gpio-cells = <2>;
+		gpio-map-mask = <0xffffffff 0xffffffc0>;
+		gpio-map-pass-thru = <0 0x3f>;		/*           shared */
+		gpio-map = <0  0 &gpio_prt10   0 0>,	/*  A0-             */
```
Comment (kartben) [inline boards/infineon/cy8ckit_062s4/cy8ckit_062s4.dts:37]: formatting of the comments is off (missing tabs?)

File: boards/infineon/cy8ckit_062s4/cy8ckit_062s4.dts [hunk 13]
```diff
@@ -29,6 +29,32 @@
...
+			   <10 0 &gpio_prt5   6 0>,	/*  D4-             */
+			   <11 0 &gpio_prt5   7 0>,	/*  D5-             */
+			   <12 0 &gpio_prt6   2 0>,	/*  D6-             */
+			   <13 0 &gpio_prt6   3 0>,	/*  D7-             */
+			   <14 0 &gpio_prt2   4 0>,	/*  D8-RX-6       y */
+			   <15 0 &gpio_prt2   6 0>,	/*  D9-TX-6       y */
+			   <16 0 &gpio_prt2   3 0>,	/* D10            y */
+			   <17 0 &gpio_prt2   0 0>,	/* D11            y */
+			   <18 0 &gpio_prt2   1 0>,	/* D12            y */
+			   <19 0 &gpio_prt2   2 0>; /* D13            y */
```
Comment (kartben) [inline boards/infineon/cy8ckit_062s4/cy8ckit_062s4.dts:56]: what are the "y"s for? looks like maybe a typo?

File: drivers/lora/rylrxxx.c [hunk 14]
```diff
@@ -0,0 +1,684 @@
...
+	msg.data = argv[3];
+	msg.rssi = atoi(argv[4]);
+	msg.snr = atoi(argv[5]);
+
+	if (RYLR_IS_RX_PENDING(driver_data->pending_async_flags)) {
+		driver_data->async_rx_cb(driver_data->dev, msg.data, msg.length, msg.rssi, msg.snr);
+	} else {
+		err = k_msgq_put(&driver_data->rx_msgq, &msg, K_NO_WAIT);
+		if (err != 0) {
+			LOG_ERR("error adding messgae to queue: %d", err);
```
Comment (kartben) [inline drivers/lora/rylrxxx.c:176]:
```suggestion
			LOG_ERR("error adding message to queue: %d", err);
```

File: drivers/lora/rylrxxx.c [hunk 15]
```diff
@@ -0,0 +1,684 @@
...
+	struct rylr_data *data = dev->data;
+
+	err = k_sem_take(&data->operation_sem, K_NO_WAIT);
+	if (err != 0) {
+		LOG_ERR("error taking operation semaphore: %d", err);
+		return err;
+	}
+
+	if (RYLR_IS_ASYNC_OP_PENDING(data->pending_async_flags)) {
+		LOG_ERR("pending async opperation");
```
Comment (kartben) [inline drivers/lora/rylrxxx.c:335]: `s/opperation/operation/g`

**Review by kartben — CHANGES_REQUESTED** [review]

**Review by kartben — APPROVED** [review]

**Review by mniestroj — APPROVED** [review]

**Comment by kartben** [comment]
@aescolar @nashif there was no change besides minor formatting/comments since assignee @JordanYates approved last so it should be ready to go. Thanks!

**Comment by aescolar** [comment]
@kartben ack, should go with the next batch

**Review by nashif — APPROVED** [review]

**Pull request merged**